edgedb-tokio/src/tutorial.md

```diff
@@ -0,0 +1,493 @@
+
+```rust
+let arguments = ("Nice movie", 2023);
+let query = "with 
+movie_title := <str>$0,
+year := <int32>$1,
+movie := (insert Movie {
+  title := movie_title,
+  release_year := year
+})
+  select  {
+    title,
+    release_year,
+    id
+}";
+let query_res: Value = client.query_required_single(query, &(arguments)).await?;
+```
+
+## The `Value` enum
+
+The [Value](https://docs.rs/edgedb-protocol/latest/edgedb_protocol/value/enum.Value.html) enum can be found in the edgedb-protocol crate. A `Value` represents anything returned from EdgeDB so you can always return a `Value` from any of the query methods without needing to deserialize into a Rust type, and the enum can be instructive in getting to know the protocol. On the other hand, returning a `Value` leads to pattern matching to get to the inner value and is not the most ergonomic way to work with results from EdgeDB.
+
+```rust
+pub enum Value {
+    Nothing,
+    Uuid(Uuid),
+    Str(String),
+    Bytes(Vec<u8>),
+    Int16(i16),
+    Int32(i32),
+    Int64(i64),
+    Float32(f32),
+    Float64(f64),
+    BigInt(BigInt),
+    // ... and so on
+}
+```
+
+Most variants of the `Value` enum correspond to a Rust type from the standard library, while some are from the `edgedb-protocol` crate and will have to be constructed. For example, this query expecting a `bigint` will return an error as it receives a `20` (an `i32`) but EdgeDB is expecting a `bigint`:
+
+```rust
+let query = "select <bigint>$0";
+let argument = 20;
+let query_res: Result<Value, _> = client.query_required_single(query, &(argument,)).await;
+assert!(format!("{query_res:?}").contains("expected std::int32"));
+```
+
+Instead, first construct a `BigInt` from the `i32` and pass that in as an argument:
+
+```rust
+use edgedb_protocol::model::BigInt;
+
+let query = "select <bigint>$0";
+let bigint_arg = BigInt::from(20);
+let query_res: Result<Value, _> = client.query_required_single(query, &(bigint_arg,)).await;
+assert_eq!(
+    format!("{query_res:?}"),
+    "Ok(BigInt(BigInt { negative: false, weight: 0, digits: [20] }))"
+);
+```
+
+## Casting inside the EdgeDB compiler
+
+EdgeDB requires arguments to have a cast in the same way that Rust requires a type declaration in function signatures. As such, arguments in queries are used as type specification for the EdgeDB compiler, not to cast from queries from the Rust side. Take this query as an example:
+
+```rust
+let query = "select <int32>$0";
+```
+
+This simply means "select an argument that must be an `int32`", not "take the received argument and cast it into an `int32`".
+
+As such, this will return an error:
+
+```rust
+let query = "select <int32>$0";
+let argument = 9i16; // Rust client will expect an int16
+let query_res: Result<Value, _> = client.query_required_single(query, &(argument,)).await;
+assert!(query_res
+    .unwrap_err()
+    .to_string()
+    .contains("expected std::int16"));
+```
+
+## Using JSON
```

Comment:
This section written as is makes it unclear of why JSON mode is needed at all. Basically it's needed to respond to the some JSON API request directly without re-serializing result. So unpacking usually isn't needed for most cases.

Reply:
Okay, added a note on JSON API requests plus an etc. at the end - sometimes people like the user on Discord last week just like the json methods for the convenience (was doing some bulk inserts)